### PR TITLE
[Proposal] Generate README.md files for all packages

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR is a proposal for generating a `README.md` file for each package in the api repo. The idea behind this PR is to make it easier to browse through the proto files from the github web UI and from preview windows in IDEs. These docs are intended only for quick reference when browsing the repo.

A few notes about the tool used:
- The tool used is a fork (in my personal space) of an existing `proto-gen-doc` tool. I had to fork it because it did not properly account for new lines in field descriptions, leading to broken tables.
- It uses html templates as the basis for the markdown conversion. This leads to some problematic conversions in some places, especially related to character escaping. This might be acceptable -- especially as a first pass.
- If this is generally useful, we can add markdown capabilities to the existing istio docgen tool.

### CODE REVIEW

File: broker/dev/README.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
+# Protocol Documentation
+<a name="top"/>
+
+## Table of Contents
+
+- [broker/dev/service_class.proto](#broker/dev/service_class.proto)
+    - [CatalogEntry](#istio.broker.dev.CatalogEntry)
```
Comment (geeknoid) [inline broker/dev/README.md]: You might need to canonicalize these IDs, . are not allowed.

File: broker/dev/README.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+- [Scalar Value Types](#scalar-value-types)
+
+
+
+<a name="broker/dev/service_class.proto"/>
+<p align="right"><a href="#top">Top</a></p>
+
+## broker/dev/service_class.proto
+Copyright 2017 Istio Authors
```
Comment (geeknoid) [inline broker/dev/README.md]: Is outputting the copyright banner intentional?

File: broker/dev/README.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  distributed under the License is distributed on an &#34;AS IS&#34; BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+
+<a name="istio.broker.dev.CatalogEntry"/>
+
+### CatalogEntry
+$hide_from_docs
```
Comment (geeknoid) [inline broker/dev/README.md]: Should either respect this, or strip it.

File: authentication/v1alpha1/README.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
+# Protocol Documentation
+<a name="top"/>
```
Comment (geeknoid) [inline authentication/v1alpha1/README.md]: What's this for?

File: broker/dev/README.md (hunk 2 above)
Comment (douglas-reid) [reply broker/dev/README.md]: Probably not. I'm guessing the source project wasn't putting copyright in their protos ;)

File: authentication/v1alpha1/README.md (hunk 4 above)
Comment (douglas-reid) [reply authentication/v1alpha1/README.md]: beats me.  Maybe something else inserted a link to `#top` somewhere?

File: broker/dev/README.md (hunk 1 above)
Comment (douglas-reid) [reply broker/dev/README.md]: ouch. ok.

Looks like the source code has too many minor issues to be to usable without modification.

Comment (geeknoid) [reply broker/dev/README.md]: proto-gen-docs has a canonicalizeID function that I use for this.

File: broker/dev/README.md (hunk 3 above)
Comment (geeknoid) [reply broker/dev/README.md]: We also have $class which is used to control CSS classes on generated objects which would need to be handled.

I think it'd be consistent to skip the element and not emit it in the docs.

File: authentication/v1alpha1/README.md (hunk 4 above)
Comment (geeknoid) [reply authentication/v1alpha1/README.md]: There are many such dummy links throughout. Weird.

**Comment by geeknoid** [comment]
Do we still want this PR?

**Comment by douglas-reid** [comment]
I think so. It just fell down the Priority ladder.

**Comment by geeknoid** [comment]
Old in the tooth, closing.

**Pull request closed without merging**